bin/configure-tower-projects.py

```diff
@@ -239,6 +240,27 @@ def extract_users(self) -> Dict[str, Users]:
             )
         return users_per_project
 
+    def extract_tags(self) -> Dict[str, Dict[str, str]]:
+        """Extract AWS tags for each stack.
+
+        Returns:
+            Mapping of projects/stacks and key-value tag pairs.
+        """
+        tags_per_project = dict()
+        for config in self.load_projects():
+            stack_name = config["stack_name"]
+            stack_tags = config.get("stack_tags", {})
+            stack_tags["TowerProject"] = stack_name
+            stack_tags.pop("OwnerEmail", None)
+            if "CostCenter" in stack_tags:
+                _, _, program_id = stack_tags["CostCenter"].rpartition("/")
+                stack_tags["CostCenter"] = program_id.strip()
+            pattern = re.compile(r"[A-z0-9_-]{1,39}")
+            assert all(pattern.fullmatch(key) for key in stack_tags.keys())
```

Comment:
If this regex fails, does that mean the tag needs to be updated upstream?

Reply:
I changed the logic to replace stretches of invalid characters with underscores instead of raising an error.